File: AdminBackend/AdminBackend/src/AdminBackend.Infrastructure/Services/Oidc/KeycloakOidcClient.cs
using AdminBackend.Application.Services.Oidc;
using IdentityModel.Client;

namespace AdminBackend.Infrastructure.Services.Oidc
{
    //NOTE : Keycloak service account must be enabled
    public class KeycloakOidcClient : IOidcService
    {
        private readonly HttpClient _httpClient;
        
        public KeycloakOidcClient(HttpClient httpClient)
        {
            _httpClient = httpClient;
        }

        public async Task<string?> GetTokenAsync(
            string authority,
            string clientId,
            string clientSecret,
            string? scope = null)
        {
            
            var discovery = await _httpClient.GetDiscoveryDocumentAsync(
                new DiscoveryDocumentRequest
                {
                    Address = $"{authority}/.well-known/openid-configuration",
                    Policy = new DiscoveryPolicy { RequireHttps = authority.StartsWith("https://") }
                });

            if (discovery.IsError)
                throw new Exception($"Discovery error: {discovery.Error}");

            var tokenResponse = await _httpClient.RequestClientCredentialsTokenAsync(
                new ClientCredentialsTokenRequest
                {
                    Address = discovery.TokenEndpoint,
                    ClientId = clientId,
                    ClientSecret = clientSecret,
                    Scope = scope
                });

            if (tokenResponse.IsError)
                throw new Exception($"Token error: {tokenResponse.Error}");

            return tokenResponse?.AccessToken;
        }
    }
}

File: DataManager/src/oidc/keycloak_oidc_client.py
# keycloak_oidc_sync.py
import httpx
from typing import Optional
#NOTE : Keycloak service account must be enabled
class KeycloakOidcClient:
    def __init__(self, client: Optional[httpx.Client] = None) -> None:        
        self.client = client or httpx.Client()

    def get_token(
        self,
        authority: str,
        client_id: str,
        client_secret: str,
        scope: Optional[str] = None,
    ) -> Optional[str]:        
        discovery_url = f"{authority}/.well-known/openid-configuration"
        resp = self.client.get(discovery_url, timeout=10)
        resp.raise_for_status()
        discovery = resp.json()

        token_endpoint = discovery.get("token_endpoint")
        if not token_endpoint:
            raise Exception("Token endpoint not found in discovery document")

        data = {
            "grant_type": "client_credentials",
            "client_id": client_id,
            "client_secret": client_secret,
        }
        if scope:
            data["scope"] = scope

        token_resp = self.client.post(token_endpoint, data=data, timeout=10)
        token_resp.raise_for_status()
        token_data = token_resp.json()

        return token_data.get("access_token")
